gh-123358: Use `_PyStackRef` in `LOAD_DEREF`

colesbury opened the pull request:
<!-- gh-issue-number: gh-123358 -->
* Issue: gh-123358
<!-- /gh-issue-number -->

corona10 committed:
gh-123358: Update LOAD_DEREF to use stackref and atomic incref

colesbury committed:
Use _Py_TryIncrefCompareStackRef and minor formatting

Fidget-Spinner commented on Include/internal/pycore_cell.h:
```diff
@@ -19,7 +20,7 @@ PyCell_SwapTakeRef(PyCellObject *cell, PyObject *value)
     Py_BEGIN_CRITICAL_SECTION(cell);
     old_value = cell->ob_ref;
-    cell->ob_ref = value;
+    FT_ATOMIC_STORE_PTR_RELEASE(cell->ob_ref, value);
```
Naiive question: out of curiosity, why does this need `release`? I don't see the `acquire` anywhere

colesbury replied:
We generally want to use at least `release` when storing pointers that may be loaded concurrently. This ensures that previously written data is visible before the store of `value` to `cell->ob_ref`.

For example, we probably initialize `value`'s type earlier in the program execution:

```
value->ob_type = &PyFloat_Type;  // for example
...
FT_ATOMIC_STORE_PTR_RELEASE(cell->ob_ref, value);
```

It's really important that value's `ob_type` field is visible *before* the write of value to `cell->ob_ref` or a reader might see some previous, garbage data for `ob_type`.

The load below uses `seq-cst`, which is at least as strong as `acquire`. The minimum in the C11/C++11 memory model would be `consume` for the load (for data dependencies), but no compiler implements that -- they all just upgrade it to "acquire", so it's kind of a mess.

Fidget-Spinner replied:
Thanks for the thorough explanation!

colesbury committed:
Merge branch 'main' into gh-123358-load-deref

colesbury committed:
Use _Py_XNewRefWithLock

colesbury committed:
Avoid slow downs in test_free_threading

With the locks around cell reads gone, some of the free threading tests
were prone to starvation: the readers were able to run in a tight loop
and the writer threads weren't scheduled frequently enough to make
timely progress.

mpage reviewed (approved):
LGTM - nice job figuring out what was going on here!
